src/main/scala/com/fulcrumgenomics/vcf/filtration/FilterSomaticVcf.scala

```diff
@@ -24,190 +24,182 @@
+      |
+      |## Performance Expectations
+      |
+      |By default `--stream-bam` will be set to false. When this tool is not asked to stream the input BAM, the BAM
+      |will be queried using random access via the BAM index file. Random access is mandatory if the input VCF is not
+      |coordinate sorted or the input BAM is not coordinate sorted as well as indexed. If the VCF is not coordinate
+      |sorted and `--stream-bam` is set to true, then an exception will be raised on the first non-coordinate increasing
+      |VCF record.
+      |
+      |Often, a VCF file will contain a sparse set of records that are scattered across a given territory within a
+      |genome or the records will be scattered genome-wide. If the territory of the VCF records is markedly smaller than
+      |the territory of all aligned SAM records in the BAM file, then random access may be the most efficient access
+      |pattern. However, there are cases where random access will be much less efficient such as when the VCF is
+      |coordinate sorted and the variant call records are very densely packed across a similar territory as compared to
+      |all aligned SAM records. Such a case is common in deeply sequenced hybrid selection NGS experiments and setting
+      |`--stream-bam` to true will often be more efficient.
+    """,
+  group = ClpGroups.VcfOrBcf
+) class FilterSomaticVcf(
+  @arg(flag = 'i', doc = "Input VCF of somatic variant calls.")         val input: PathToVcf,
+  @arg(flag = 'o', doc = "Output VCF of filtered somatic variants.")    val output: PathToVcf,
+  @arg(flag = 'b', doc = "BAM file for the tumor sample.")              val bam: PathToBam,
+  @arg(flag = 's', doc = "Sample name in VCF if `> 1` sample present.") val sample: Option[String]   = None,
+  @arg(flag = 'm', doc = "Minimum mapping quality for reads.")          val minMappingQuality: Int   = 20,
+  @arg(flag = 'q', doc = "Minimum base quality.")                       val minBaseQuality: Int      = 20,
+  @arg(flag = 'p', doc = "Use only paired reads mapped in pairs.")      val pairedReadsOnly: Boolean = false,
+  @arg(flag = 'R', doc = "Whether to stream the BAM input or use BAM random access.") val streamBam: Boolean = false,
+  @arg(doc = "Distance from 5-prime end of read to implicate A-base addition artifacts. Set to :none: to deactivate the filter.")
+  val aTailingDistance: Option[Int] = Some(2),
+  @arg(doc = "Minimum acceptable p-value for the A-base addition artifact test.")
+  val aTailingPValue: Option[Double] = None,
+  @arg(doc = "Distance from 5-prime end of read to implicate end repair fill-in artifacts. Set to :none: to deactivate the filter.")
+  val endRepairFillInDistance: Option[Int] = Some(15),
+  @arg(doc = "Minimum acceptable p-value for the end repair fill-in artifact test.")
+  val endRepairFillInPValue: Option[Double] = None
 ) extends FgBioTool with LazyLogging {
-  Io.assertReadable(input)
-  Io.assertReadable(bam)
-  Io.assertCanWriteFile(output)
```

Comment:
Two reasons mostly:

1. I've been bit by race conditions this style of validation causes when placing checks in class constructors. This can often happen when using these tools programmatically. Imagine a case where you want to call this tool programmatically so you build the class but you have not had a chance to build the input files before calling `.execute()`.
2. The handles opened over these resources all raise exceptions immediately that explain the issue (`AccessDeniedException` or `PermissionDenied`).

I can bring them back, but I would prefer moving them to the first statement in the `.execute()` block so validation runs immediately before needing the files, minimizing the chance for race conditions (it's better to ask for forgiveness vs permission).

Raises a good point though. How come the VCF/BAM source and writers don't perform this check themselves?

Reply:
Let's move them back as this is a command line tool and not meant to be called programatically.  If you want to use it programmatically, we should discuss how in more detail offline about how to refactor.

Having them in the constructor will yield a nicer formatted usage error message from `sopt` [here](https://github.com/fulcrumgenomics/fgbio/blob/c26d405c55872b683dfa16ffe8d3c360db5ce898/src/main/scala/com/fulcrumgenomics/cmdline/FgBioMain.scala#L102) whereas it is less clear that it's a command line argument when thrown from the usage.